Kernel: Add NVMe support to serenity

Panky-codes opened the pull request:
Add a basic NVMe support to do RW in serenity based on NVMe spec [1.4 ](https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf).
### Implementation
Multiple NVMe drives(subsystems) can be connected at the same time but the implementation supports only one NVMe controller with multiple namespaces per NVMe drive.  

For smp based systems, each core gets a unique IO Queue. As the OS lacks MSI support, pin based interrupts are used to indicate request completion. 

The driver was tested by replacing the current BMIDE driver with the NVMe driver :^) .
### Usage
To enable NVMe support, run the following:

```    
$ SERENITY_NVME_ENABLE=1 Meta/serenity.sh run i686 root=/dev/nvme0n1
```
![image](https://user-images.githubusercontent.com/33182938/145698495-a1e6d75a-31c8-4766-a5c2-089acfc2a5ec.png)

### Limitation
One major limitation of this initial implementation is it processes **one request at a time** and supports a **maximum of 4k IO transfers**. These constraints also holds true to the current BMIDE driver as these are enforced in the block layer. But the plan is to eventually remove them in the NVMe driver so that we can process multiple request at a time and support IO requests that are bigger than 4k. 
 
### Open issues
- For some reason `/sys/dev/block` doesn't show a device with major number 3, even though Storage device is constructed with major number and minor number from `StorageManagement class`:
![image](https://user-images.githubusercontent.com/33182938/145698522-2b971375-66d5-4483-9112-66d13f4ac85b.png)
    - Fixed by using DeviceManagement API

Sylvyrfysh commented on Kernel/Devices/BlockDevice.h:
```diff
@@ -68,6 +68,8 @@ class BlockDevice : public Device {
         , m_block_size(block_size)
     {
+        // 512 is the minimum sector size in most block devices
+        VERIFY(m_block_size >= 512);
```
Any particular reason this needs to be here? If we have no code that needs this, should this be a (pointless) requirement?

Panky-codes replied:
I spent/wasted a lot of time because initially as I was passing a wrong `m_block_size` which lead to a memory corruption error. I thought it would be a good check to add to avoid future misery. 
Anyway, almost all storage devices has 512 bytes as a minimum block size, that is why added this check.

Sylvyrfysh commented on Kernel/Storage/NVME/NVMEController.cpp:
```diff
@@ -0,0 +1,315 @@
+    , m_admin_queue_ready(false)
+    , m_device_count(0)
+{
+    // Nr of queues = one queue per core
```
Is this common? Does Linux do something similar, or multiple per core? If not, is this just much simpler to implement?

Panky-codes replied:
Yeah, Linux also does this. I don't think it is anymore complicated by constructing a queue per core, but the wins are huge. 
Currently the benefits can't be fully utilized in a `smp` system because we are using PIN based interrupts. Once we start supporting MSI interrupts and port NVMe code to use an unique interrupt vector per queue, then it will enable parallel processing without any contention on IO queues.

Sylvyrfysh commented on Kernel/Storage/NVME/NVMEController.cpp:
```diff
@@ -0,0 +1,315 @@
+
+    return true;
+}
+u32 NVMEController::get_admin_q_dept()
```
Inconsistent codestyle: add newlines between functions

Sylvyrfysh commented on Kernel/Storage/NVME/NVMEController.cpp:
```diff
@@ -0,0 +1,315 @@
+}
+Tuple<u64, u8> NVMEController::get_ns_features(ByteBuffer& identify_data_struct)
+{
+    auto flbas = identify_data_struct[26] & 0xf;
```
I see a lot of magic numbers around here, either adding comments to the spec or pulling them out into constants and commenting those at the top of the file I would consider useful

Panky-codes replied:
Thanks for the comment. I tried to avoid them but some slipped through. I will add these as constants in the next revision.

Sylvyrfysh commented on Kernel/Storage/NVME/NVMEController.h:
```diff
@@ -0,0 +1,85 @@
+        return *reinterpret_cast<T*>(ptr + offset);
+    }
+
+public:
```
Any reason for the multiple public/private access qualifiers?

Sylvyrfysh commented on Meta/run.sh:
```diff
@@ -183,6 +183,17 @@ else
+    SERENITY_NVME_ARGS="$SERENITY_NVME_ARGS -device i82801b11-bridge,id=bridge4 -device sdhci-pci,bus=bridge4"
+    SERENITY_NVME_ARGS="$SERENITY_NVME_ARGS -device nvme,serial=deadbeef,drive=disk,bus=bridge4"
+else
+    SERENITY_NVME_ARGS="-drive file=${SERENITY_DISK_IMAGE},format=raw,index=0,media=disk"
```
Rename to SERENITY_BOOT_DRIVE?

Panky-codes replied:
This is definitely a better name than `SERENITY_NVME_ARGS`. Will use this in my next revision.

Sylvyrfysh reviewed (commented):
Looks pretty good overall. Biggest thing would be the VERIFY of 512 byte blocks, I'm not the expert on codestyle by any means.

ADKaster reviewed (commented):
This looks like a pretty decent start to some NVME features in the kernel! Few comments: some style guide related issues, and some memory allocation questions.

I think the answer to my "why are you VERIFY()-ing that allocation succeeds here" is likely because the code was heavily inspired by the existing storage device drivers. If that's the case then maybe I should direct those questions to an issue and tag @supercomputer7  instead :).

Panky-codes commented on Kernel/Storage/NVME/NVMEController.cpp:
```diff
@@ -0,0 +1,315 @@
+
+NonnullRefPtr<NVMEController> NVMEController::initialize(const Kernel::PCI::DeviceIdentifier& device_identifier)
+{
+    return adopt_ref(*new NVMEController(device_identifier));
```
Thanks for the comment! I tried to keep the pattern consistent across storage devices. But I totally agree with your point. I would like to hear @supercomputer7 view as well. 
But I will try to take your comments into account in my next revision.

supercomputer7 replied:
@ADKaster is right here, we should write new kernel code with OOM mindset. It will take time to propagate these patterns but eventually we will be able to handle OOM much better than today :)

Panky-codes commented on Kernel/Storage/NVME/NVMEQueue.h:
```diff
@@ -0,0 +1,77 @@
+class AsyncBlockDeviceRequest;
+class NVMEQueue : public IRQHandler
+    , public RefCounted<NVMEQueue> {
+    AK_MAKE_ETERNAL
```
Honestly, I tried different things to get it up and running and I finally forgot to remove it. Thanks for pointing out!

Panky-codes replied:
I also noticed a similar pattern in StorageController class: 
```
class StorageController : public RefCounted<StorageController> {
    AK_MAKE_ETERNAL
```
I think `AK_MAKE_ETERNAL` should be removed there as well.

Panky-codes commented on Kernel/Storage/NVME/NVMENameSpace.cpp:
```diff
@@ -0,0 +1,109 @@
+        queue.write(request, m_nsid, request.block_index(), request.block_count());
+    }
+}
+void NVMENameSpace::test_read()
```
Exactly. I used it extensively to test my implementation but currently there are no users. I will remove it for now. 

Built-in self tests like kselftests in Linux sounds like a good idea.

supercomputer7 replied:
While I agree on the most of this comment, I highly doubt it is a good idea to create one unified interface for built-in self test for hardware components in the Kernel. However, for NVMe, we can use a known character device for built-in self test and also to expand it to manage the controller like what has been done in Linux (e.g. by using `/dev/nvmeX` device).

supercomputer7 commented on Kernel/Storage/NVME/NVMEDefinitions.h:
```diff
@@ -0,0 +1,102 @@
+#define OP_ADMIN_CREATE_COMPLETION_QUEUE 0x5
+#define OP_ADMIN_CREATE_SUBMISSION_QUEUE 0x1
+#define OP_ADMIN_IDENTIFY 0x6
+#define OP_ADMIN_SET_FEATURES 0x9
```
I personally prefer enum classes in a known namespace, but that's not my work here. Even enums are better than plain macros, but it's up to you to decide what to do with this.

Panky-codes replied:
I will use `constexpr` / `enum class` here instead of macros. I use `C` a lot and sometimes old habits creep in ;)

supercomputer7 commented on Kernel/Storage/NVME/NVMENameSpace.cpp:
```diff
@@ -0,0 +1,107 @@
+    // TODO: For now we are assuming only one controller is there. Name format nvme<ctrlid>n<nsid>
+    auto device_name = String::formatted("nvme0n{:d}", nsid);
+    auto device_name_kstring = KString::must_create(device_name.view());
+    return adopt_ref(*new NVMENameSpace(queues, storage_size, lba_size, major_number, minor_number, nsid, move(device_name_kstring)));
```
This is not the correct way to create a block/character device.
You should do something like this:
```c++
    auto disk_device_or_error = DeviceManagement::try_create_device<NVMENameSpace>(...);
```
This is why the disk device doesn't appear on the `/sys/dev/block` directory.

supercomputer7 commented:
Nitpick: On the Internet, NVM Express is usually short-named to "NVMe" and not "NVME".
I'd highly appreciate it if you could stick to the first name ("NVMe") for all proposed code.

Sylvyrfysh commented on Meta/run.sh:
```diff
@@ -213,6 +223,8 @@ if [ -z "$SERENITY_MACHINE" ]; then
         -device i82801b11-bridge,id=bridge3 -device sdhci-pci,bus=bridge3
         -device ich9-ahci,bus=bridge3
         -chardev stdio,id=stdout,mux=on
+        -qmp unix:qmp-sock,server,nowait
```
This was removed by #11237 , consider rebasing on master or removing. Looks good otherwise, thanks for the extra comments!

Panky-codes replied:
Good catch! Probably I did something funny during my rebase!
Anyway, I have to do a big refactoring/rearranging based on @ADKaster's comment.
I will include it then.

Panky-codes commented:
Good catch! Probably I did something funny during my rebase!
Anyway, I have to do a big refactoring/rearranging based on @ADKaster's comment.
I will include it then. 
On Mon, Dec 13, 2021 at 08:20:21PM -0800, Nick Johnson wrote:
> @Sylvyrfysh commented on this pull request.
> 
> 
> 
> > @@ -213,6 +223,8 @@ if [ -z "$SERENITY_MACHINE" ]; then
>          -device i82801b11-bridge,id=bridge3 -device sdhci-pci,bus=bridge3
>          -device ich9-ahci,bus=bridge3
>          -chardev stdio,id=stdout,mux=on
> +        -qmp unix:qmp-sock,server,nowait
> 
> This was removed by #11237 , consider rebasing on master or removing. Looks good otherwise, thanks for the extra comments!
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SerenityOS/serenity/pull/11222#pullrequestreview-831000740

Panky-codes commented:
Good catch! Probably I did something funny during my rebase!
Anyway, I have to do a big refactoring/rearranging based on @ADKaster's comment.
I will include it then.

Panky-codes commented:
Ping @supercomputer7 & @ADKaster

supercomputer7 commented:
I've written my requests to you on Discord now, but I'll include them here:

> 9:09 PM] supercomputer7: I read somewhere that you actually have to not refer to the NVMe spec if you want to implement it because the spec "doesnt allow it". regardless of that, besides that I do think there's more work to do on this code but I can't expect it to be perfect right now :)
[9:09 PM] supercomputer7: I don't want us to be able to attach the driver on only one NVMe controller
[9:10 PM] supercomputer7: so please ensure this doesn't happen
[9:10 PM] supercomputer7: everything else can be improved afterwards :)

Panky-codes commented:
> > 9:09 PM] supercomputer7: I read somewhere that you actually have to not refer to the NVMe spec if you want to implement it because the spec "doesnt allow it". regardless of that, besides that I do think there's more work to do on this code but I can't expect it to be perfect right now :)
Oh, I didn't know this. I removed all the reference to NVMe Spec now.
> [9:09 PM] supercomputer7: I don't want us to be able to attach the driver on only one NVMe controller
Yeah, we will now be able to connect multiple NVMe drives(susbystems) without any
name clashes in the drive name :)

But sorry if I confused you in my description about one Controller. 
I was talking about NVMe Spec's multi-controller, shared namespace drives,
which we don't support currently. I hope I am clear now :)
> [9:10 PM] supercomputer7: so please ensure this doesn't happen
> [9:10 PM] supercomputer7: everything else can be improved afterwards :)
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/SerenityOS/serenity/pull/11222#issuecomment-996102330
> You are receiving this because you authored the thread.
> 
> Message ID: ***@***.***>

Panky-codes commented:
> > 9:09 PM] supercomputer7: I read somewhere that you actually have to not refer to the NVMe spec if you want to implement it because the spec "doesnt allow it". regardless of that, besides that I do think there's more work to do on this code but I can't expect it to be perfect right now :)

Oh, I didn't know this. I removed all the reference to NVMe Spec now.

> [9:09 PM] supercomputer7: I don't want us to be able to attach the driver on only one NVMe controller

Yeah, we will now be able to connect multiple NVMe drives(susbystems) without any
name clashes in the drive name :)

As you can see in the image below where two NVMe drives are connected:
![Screenshot from 2021-12-17 10-40-14](https://user-images.githubusercontent.com/33182938/146494053-10be57e3-1a64-48de-8b99-ce420391e732.png)

But sorry if I confused you in my description about one Controller. 
I was talking about NVMe Spec's multi-controller, shared namespace drives,
which we don't support currently. I hope I am clear now :)
> [9:10 PM] supercomputer7: so please ensure this doesn't happen
> [9:10 PM] supercomputer7: everything else can be improved afterwards :)

supercomputer7 commented on Kernel/Memory/MemoryManager.cpp:
```diff
@@ -714,6 +714,15 @@ ErrorOr<NonnullOwnPtr<Region>> MemoryManager::allocate_contiguous_kernel_region(
     return allocate_kernel_region_with_vmobject(range, move(vmobject), name, access, cacheable);
 }
 
+ErrorOr<NonnullOwnPtr<Memory::Region>> MemoryManager::dma_allocate_buffer(size_t size, StringView name, Memory::Region::Access access, RefPtr<Memory::PhysicalPage>& dma_buffer_page)
```
I would call this `allocate_dma_buffer` instead.

supercomputer7 commented on Kernel/Storage/NVMe/NVMeQueue.h:
```diff
@@ -0,0 +1,77 @@
+    u8 cq_valid_phase;
+    u16 sq_tail;
+    u16 prev_sq_tail {};
+    u16 cq_head;
```
Please add the `m_` prefix to these class members :)

supercomputer7 commented on Kernel/Storage/NVMe/NVMeNameSpace.cpp:
```diff
@@ -0,0 +1,49 @@
+    : StorageDevice(major_number, minor_number, lba_size, max_addresable_block, move(dev_name))
+    , m_nsid(nsid)
+    , m_lba_size(lba_size)
+    , m_max_addresable_block(max_addresable_block)
```
Do we actually need this? StorageDevice already has such member.

Panky-codes replied:
Good point! Also removed lba_size because block device has that member!

supercomputer7 commented on Kernel/Storage/NVMe/NVMeQueue.h:
```diff
@@ -0,0 +1,77 @@
+    RefPtr<Memory::PhysicalPage> m_sq_dma_page;
+    Span<NVMeCompletion> m_cqe_array;
+    OwnPtr<Memory::Region> m_rw_dma_region;
+    Memory::Region* m_db_regs;
```
We should probably use a OwnPtr instead of naked raw pointer :)

supercomputer7 commented on Kernel/Storage/NVMe/NVMeQueue.h:
```diff
@@ -0,0 +1,77 @@
+public:
+    static ErrorOr<NonnullRefPtr<NVMeQueue>> try_create(u16 qid, u8 irq, u32 q_depth, OwnPtr<Memory::Region> cq_dma_region, RefPtr<Memory::PhysicalPage> cq_dma_page, OwnPtr<Memory::Region> sq_dma_region, RefPtr<Memory::PhysicalPage> sq_dma_page, Memory::Region* db_regs);
+    ErrorOr<void> create();
+    explicit NVMeQueue(u16 qid, u8 irq, u32 q_depth, OwnPtr<Memory::Region> cq_dma_region, RefPtr<Memory::PhysicalPage> cq_dma_page, OwnPtr<Memory::Region> sq_dma_region, RefPtr<Memory::PhysicalPage> sq_dma_page, Memory::Region* db_regs);
```
Technically we should use OwnPtr Here too so Memory::Region* db_regs should be OwnPtr<Memory::Region>
and you should really use move semantics here :)

supercomputer7 commented on Kernel/Storage/NVMe/NVMeController.cpp:
```diff
@@ -0,0 +1,318 @@
+    auto bar_len = round_up_to_power_of_two(CTRL_REG_SIZE(nr_of_queues), 4096);
+    auto region_or_error = TRY(MM.allocate_kernel_region(PhysicalAddress(bar), bar_len, "PCI NVMe BAR", Memory::Region::Access::ReadWrite));
+    // Queues need access to the doorbell registers, hence leaking the pointer
+    m_controller_regs = region_or_error.leak_ptr();
```
Please don't leak a pointer here. Maybe using `TypedMapping` struct more could help you to do this more cleanly :)

supercomputer7 commented:
This is really nice work you did here. I told you we need a few more fixes and from my opinion it's good to go, but it'd be awesome if you could leverage more the `TypedMapping` struct here to ensure we don't need to leak pointers or do the pointer arithmetic ourselves :)

Panky-codes commented:
Thanks for the `TypedMapping` tip. It helped me from not leaking the pointer :) I also made some more cleanup to the code after addressing your comments! 

Cheers!

supercomputer7 commented on Kernel/Storage/NVMe/NVMeController.h:
```diff
@@ -0,0 +1,91 @@
+    RefPtr<NVMeQueue> m_admin_queue;
+    NonnullRefPtrVector<NVMeQueue> m_queues;
+    NonnullRefPtrVector<NVMeNameSpace> m_namespaces;
+    Memory::TypedMapping<u8> m_controller_regs;
```
We could put it this way for this patch and it will be OK for now, but later on we will need to write the structure definition for this mapping instead. This way we don't need to do "ugly" `reinterpret_cast` to access registers with the required bit width access each time, but to rely on the structure definition and the compiler to do this for us.
Whether you want to handle it now or later is completely up to you :)

Panky-codes replied:
That is actually a good idea! :) I got rid of most of the `reinterpret_cast`. :)
Of course there are some more but I don't think those are required because we need to write those pointers to registers. But eventually we can also see if we can get rid of them :)

Panky-codes commented:
Ping @supercomputer7

supercomputer7 reviewed (approved):
Great work :)

FireFox317 commented on Meta/run.sh:
```diff
@@ -183,6 +183,17 @@ else
+# FIXME: Is there a better way of enabling NVME support?
+SERINITY_NVME_ENABLE=0
+if [ "$SERINITY_NVME_ENABLE" -eq 1 ]; then
+
```
One nit: why is there an empty line here?

Panky-codes replied:
Just bad eyesight! :P Fixed it! Thanks

Panky-codes commented:
Thanks for your help @supercomputer7 . The code is much cleaner now! :)

Panky-codes commented:
@awesomekling Could you merge this PR when you time?

awesomekling commented on Kernel/Memory/MemoryManager.cpp:
```diff
@@ -725,6 +725,15 @@ ErrorOr<NonnullOwnPtr<Region>> MemoryManager::allocate_contiguous_kernel_region(
     return allocate_kernel_region_with_vmobject(range, move(vmobject), name, access, cacheable);
 }
 
+ErrorOr<NonnullOwnPtr<Memory::Region>> MemoryManager::allocate_dma_buffer(size_t size, StringView name, Memory::Region::Access access, RefPtr<Memory::PhysicalPage>& dma_buffer_page)
```
Hmm, this API is a bit strange. What is supposed to happen if I pass a `size` larger than `PAGE_SIZE`?

Panky-codes replied:
Good catch.
I added two APIs instead of one now: 
```cpp
ErrorOr<NonnullOwnPtr<Memory::Region>> allocate_dma_buffer_page(StringView name, Memory::Region::Access access, RefPtr<Memory::PhysicalPage>& dma_buffer_page);
ErrorOr<NonnullOwnPtr<Memory::Region>> allocate_dma_buffer_pages(size_t size, StringView name, Memory::Region::Access access, NonnullRefPtrVector<Memory::PhysicalPage>& dma_buffer_pages);
```
I hope the API makes sense now.

awesomekling commented on Kernel/Storage/NVMe/NVMeQueue.h:
```diff
@@ -0,0 +1,81 @@
+    }
+
+private:
+    u16 m_qid;
+    u8 m_cq_valid_phase;
+    u16 m_sq_tail;
+    u16 m_prev_sq_tail {};
+    u16 m_cq_head;
```
Please explicitly initialize primitive variables.

awesomekling commented:
Please wrap commit messages at 72 characters (per the [contributing guidelines](https://github.com/SerenityOS/serenity/blob/9eb08bdb0f7460eb9874a5d16f6f3808f3d8a8f0/CONTRIBUTING.md)).

Panky-codes commented:
@awesomekling Addressed your comments. Wrt `wrap commit message` comment, I think it was wrapped before as well. Or else the commit message linter would have also complained right?

Panky-codes commented:
Ping @awesomekling

supercomputer7 commented:
Congrats! :)
